PR: https://github.com/odoo/odoo/pull/

From: 99f497b390e680629dce3a7e2e282ada74ee9884
From: Yannick Tivisse
Date: 2018-04-26 13:13:38

Structural Changes: 6
Total Changes: 94

[IMP] *: Define groups on res.users models

The reified view on the res users will be dropped in the following commit.

The previous commit adds support to define each group as a computed field on the res users.

This commit defines:
- A boolean field for each 'isolated' res.group, i.e. a group in the hidden category.
- A selection field for each 'Application' res.group, i.e. a group in a application category.

Example:
- The group to manage pricelist in sales becomes a boolean field
- The groups project user/manager become a selection field

================================= pseudo patch: =================================

--- a/odoo/addons/base/data/ir_module_category_data.xml
+++ b/odoo/addons/base/data/ir_module_category_data.xml
@@ -171,11 +171,6 @@
             <field name="sequence">101</field>
         </record>
 
-        <record model="ir.module.category" id="module_category_extra">
-            <field name="name">Other Extra Rights</field>
-            <field name="sequence">102</field>
-        </record>
-
         <!-- add applications to base groups -->
         <record model="res.groups" id="group_erp_manager">
             <field name="category_id" ref="module_category_administration"/>

--- a/odoo/addons/base/data/res_users_demo.xml
+++ b/odoo/addons/base/data/res_users_demo.xml
@@ -72,6 +72,9 @@
             <field name="groups_id" eval="[(5,)]"/><!-- Avoid auto-including this user in any default group -->
         </record>
 
+        <record id="base.group_user" model="res.groups"><!-- Remove the demo user for the internal_user group -->
+            <field name="users" eval="[(3,ref('demo_user0'))]"/>
+        </record>
         <record id="base.group_portal" model="res.groups"><!-- Add the demo user to the portal (and therefore to the portal member group) -->
             <field name="users" eval="[(4,ref('demo_user0'))]"/>
         </record>

--- a/odoo/addons/base/models/res_users.py
+++ b/odoo/addons/base/models/res_users.py
@@ -230,6 +230,32 @@ class Users(models.Model):
     name = fields.Char(related='partner_id.name', inherited=True)
     email = fields.Char(related='partner_id.email', inherited=True)
 
+    group_base_user = fields.Selection(
+        selection=lambda self: self._get_group_selection('base.module_category_user_type'),
+        string="User type", compute='_compute_groups_id', inverse='_inverse_groups_id',
+        category_xml_id='base.module_category_user_type')
+
+    group_administration_user = fields.Selection(
+        selection=lambda self: self._get_group_selection('base.module_category_administration'),
+        string="Administration", compute='_compute_groups_id', inverse='_inverse_groups_id',
+        category_xml_id='base.module_category_administration')
+
+    has_group_multi_company = fields.Boolean(
+        'Multi Companies', compute='_compute_groups_id', inverse='_inverse_groups_id',
+        group_xml_id='base.group_multi_company')
+
+    has_group_multi_currency = fields.Boolean(
+        'Multi Currencies', compute='_compute_groups_id', inverse='_inverse_groups_id',
+        group_xml_id='base.group_multi_currency')
+
+    has_group_no_one = fields.Boolean(
+        'Technical Features', compute='_compute_groups_id', inverse='_inverse_groups_id',
+        group_xml_id='base.group_no_one')
+
+    has_group_partner_manager = fields.Boolean(
+        'Contact Creation', compute='_compute_groups_id', inverse='_inverse_groups_id',
+        group_xml_id='base.group_partner_manager')
+
     _sql_constraints = [
         ('login_key', 'UNIQUE (login)',  'You can not have two users with the same login !')
     ]
@@ -330,6 +356,7 @@ class Users(models.Model):
 
     @api.model
     def create(self, vals):
+        # import pudb; pudb.set_trace()
         user = super(Users, self).create(vals)
         user.partner_id.active = user.active
         if user.partner_id.company_id:
@@ -766,6 +793,7 @@ class GroupsImplied(models.Model):
     @api.multi
     def write(self, values):
         res = super(GroupsImplied, self).write(values)
+
         if values.get('users') or values.get('implied_ids'):
             # add all implied groups (to all users of each group)
             for group in self:

--- a/odoo/addons/base/views/res_users_views.xml
+++ b/odoo/addons/base/views/res_users_views.xml
@@ -141,11 +141,6 @@
                             <field name="company_id" context="{'user_preference': 0}" groups="base.group_multi_company"/>
                         </div>
                         <group>
-                            <label for="groups_id" string="Access Rights"
-                                    attrs="{'invisible': [('id', '>', 0)]}" groups="base.group_no_one"/>
-                            <div attrs="{'invisible': [('id', '>', 0)]}" groups="base.group_no_one">
-                                <field name="groups_id" readonly="1" widget="many2many_tags" options="{'color_field': 'color'}" style="display: inline;"/> You will be able to define additional access rights by editing the newly created user under the Settings / Users menu.
-                            </div>
                             <field name="phone" widget="phone"/>
                             <field name="mobile" widget="phone"/>
                         </group>
@@ -181,12 +176,36 @@
                         </div>
                         <notebook colspan="4">
                             <page name="access_rights" string="Access Rights">
+                                <group string="User Type" name="user_type" groups="base.group_no_one">
+                                    <field name="group_base_user" groups="base.group_no_one"/>
+                                </group>
                                 <group string="Multi Companies" attrs="{'invisible': [('companies_count', '&lt;=', 1)]}">
                                     <field string="Allowed Companies" name="company_ids" widget="many2many_tags" options="{'no_create': True}"/>
                                     <field string="Current Company" name="company_id" context="{'user_preference': 0}"/>
                                     <field string="Companies count" name="companies_count" invisible="1"/>
                                 </group>
-                                <field name="groups_id"/>
+                                <group string="Application Accesses" name="applications"/>
+                                <group string="Technical Settings" name="technical" groups="base.group_no_one" col="8">
+                                    <!-- TODO: Hide the no_one group -->
+                                    <label for="has_group_no_one" colspan="3"/>
+                                    <div>
+                                        <field name='has_group_no_one' colspan="1" class="oe_inline"/>
+                                    </div>
+                                </group>
+                                <group string="Extra Rights" name="extra" groups="base.group_no_one" col="8">
+                                    <label for="has_group_multi_company" colspan="3"/>
+                                    <div>
+                                        <field name='has_group_multi_company' colspan="1" class="oe_inline"/>
+                                    </div>
+                                    <label for="has_group_multi_currency" colspan="3"/>
+                                    <div>
+                                        <field name='has_group_multi_currency' colspan="1" class="oe_inline"/>
+                                    </div>
+                                    <label for="has_group_partner_manager" colspan="3"/>
+                                    <div>
+                                        <field name='has_group_partner_manager' colspan="1" class="oe_inline"/>
+                                    </div>
+                                </group>
                             </page>
                             <page string="Preferences">
                                 <group>
@@ -203,12 +222,27 @@
                                     <field name="signature"/>
                                 </group>
                             </page>
+                            <page string="Groups" groups="base.group_no_one">
+                                <group>
+                                    <field name="groups_id"/>
+                                </group>
+                            </page>
                         </notebook>
                     </sheet>
                 </form>
             </field>
         </record>
-
+        <record id="res_users_view_form_administation" model="ir.ui.view">
+            <field name="name">res.users.view.form.administration</field>
+            <field name="model">res.users</field>
+            <field name="inherit_id" ref="view_users_form"/>
+            <field name="priority">1000</field>
+            <field name="arch" type="xml">
+                <group name="applications" position="inside">
+                    <field name="group_administration_user"/>
+                </group>
+            </field>
+        </record>
         <record id="view_users_tree" model="ir.ui.view">
             <field name="name">res.users.tree</field>
             <field name="model">res.users</field>
@@ -268,16 +302,6 @@
                 </search>
             </field>
         </record>
-        <record id="user_groups_view" model="ir.ui.view">
-            <field name="name">res.users.groups</field>
-            <field name="model">res.users</field>
-            <field name="inherit_id" ref="view_users_form"/>
-            <field name="arch" type="xml">
-                <!-- dummy, will be modified by groups -->
-                <field name="groups_id" position="after"/>
-            </field>
-        </record>
-        <!-- dummy update on group, to force the view above to be update -->
         <record model="res.groups" id="group_no_one"/>
         <record id="action_res_users" model="ir.actions.act_window">
             <field name="name">Users</field>
